code/go/0chain.net/chaincore/chain/state.go

```diff
@@ -177,7 +177,9 @@ func (c *Chain) updateState(ctx context.Context, b *block.Block, txn *transactio
 		if output, err = c.ExecuteSmartContract(ctx, txn, sctx); err != nil {
 			logging.Logger.Error("Error executing the SC", zap.Any("txn", txn),
 				zap.Error(err))
-			return
+			txn.TransactionOutput = err.Error()
+			txn.Status = transaction.TxnFail
+			return nil
```

Comment:
Please do not ignore the returned error if the transaction execute failed.

Reply:
It was the intention that failed smart contract transactions not be considered internal failures and instead become part of the block. If an error is returned, a transaction will be ignored.
Could you suggest some unintended consequences if you're aware of something?

Reply:
Transactions could fail for different reasons, and we can not ignore all of the transactions directly. Check the following situations:
a) The transactions are invalid by themself, we should and must ignore them (The question is whether this transactions could be packed into a block, need further check)
b) The transactions are valid, and it could be executed successfully by most of the miners and sharders, except some miners that do not have sufficient state in MPT. In this situation, if we ignore the transaction, and continue the block state compute, the block's MPT state would be different from others, and we will see the `mismatch  state hash` error. What we should do is return, instead of continue the block state computing.

Reply:
So basically, we need to differ the transaction fail errors. 
a) invalid transactions by themself
b) Transactions are valid, but the miner is not ready yet to execute the transactions.

Reply:
Might be a dumb question, but could we check the transaction validity when the client submits it? Then it'd be no surprise for the txn to be discarded and not included in the block.

Reply:
valid point by @peterlimg.I can add a little bit, there can be 3 main cases:
1. transaction is malformed or has invalid signature or anything else that can be validated at the time of receiving
2. transaction can't be executed now, but can be executed in the future (no funds yet for transfer and so on) not sure that there are lots of cases, since we use transaction creation time for sequencing.
3. transaction causes error to smart contract 

1. should not be included into the block and should be excluded from tx_pool asap, we have to collect statistics on malicious senders by ip or something similar
2. should be stored in the tx_pool for some time and if can't be executed successfully should be included into the block
3. should be included into the block